cpp/src/arrow/dataset/scanner.cc

```diff
@@ -224,5 +225,100 @@ Result<std::shared_ptr<Table>> Scanner::ToTable() {
                                   FlattenRecordBatchVector(std::move(state->batches)));
 }
 
+struct ToBatchesState {
+  explicit ToBatchesState(size_t n_tasks)
+      : batches(n_tasks), task_drained(n_tasks, false) {}
+
+  /// Protecting mutating accesses to batches
+  std::mutex mutex;
+  std::vector<std::deque<std::shared_ptr<RecordBatch>>> batches;
+  std::vector<bool> task_drained;
+  size_t pop_cursor = 0;
+
+  void Push(std::shared_ptr<RecordBatch> b, size_t i_task) {
+    std::lock_guard<std::mutex> lock(mutex);
+    if (batches.size() <= i_task) {
+      batches.resize(i_task + 1);
+      task_drained.resize(i_task + 1);
+    }
+    batches[i_task].push_back(std::move(b));
+  }
+
+  Status Finish(size_t position) {
+    std::lock_guard<std::mutex> lock(mutex);
+    task_drained[position] = true;
+    return Status::OK();
+  }
+
+  std::shared_ptr<RecordBatch> Pop() {
+    std::unique_lock<std::mutex> lock(mutex);
+    std::condition_variable().wait_for(lock, std::chrono::milliseconds{1}, [this] {
+      while (pop_cursor < batches.size()) {
+        // queue for current scan task contains at least one batch, pop that
+        if (!batches[pop_cursor].empty()) return true;
+
+        // queue is empty but will be appended to eventually, wait for that
+        if (!task_drained[pop_cursor]) return false;
+
+        ++pop_cursor;
+      }
+      // all scan tasks drained, terminate
+      return true;
+    });
+
+    if (pop_cursor == batches.size()) return nullptr;
+
+    auto batch = std::move(batches[pop_cursor].front());
+    batches[pop_cursor].pop_front();
+    return batch;
+  }
+};
+
+Result<RecordBatchIterator> Scanner::ToBatches() {
+  ARROW_ASSIGN_OR_RAISE(auto scan_task_it, Scan());
+  ARROW_ASSIGN_OR_RAISE(auto scan_task_vector, scan_task_it.ToVector());
```

Comment:
I'll be replacing this with something better so I don't know how much we care to worry but this is not ideal.  For example, with parquet, this would fetch metadata for every file in the scan before starting to read any individual file.  It introduces more latency than necessary.

Also, I'm not sure how this will interact with parquet preloading.